### PR TITLE
LPD-20528: Return default response value locale when user locale is not registered in each Liferay instance

### PR DESCRIPTION
**Pull request by sergiojimcos**
## Motivation
The idea of this PR is to return the default user locale when an user request a Liferay Headless API and the locate is not registered in the company. Before this PR, this will throw a `NotAcceptableException`. This only is going to be applied to GET requests.

**What is new?**
- Return the default user locale registered if the locale requested is not registered in the company.
- Update test case to assert that `NotAcceptableException` is not thrown.

**Steps to reproduce**
1. Execute this curl:

```curl
curl -X 'GET' 'http://localhost:8080/o/headless-admin-user/v1.0/roles' -H 'accept: application/json' -u test@liferay.com:test -H 'Accept-Language: vi-VN, vi;q=0.9'
```
**JIRA Related Ticket**
https://liferay.atlassian.net/browse/LPD-20528

### CODE REVIEW

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Review by ccorreagg — COMMENTED** [review]
Please review my comments

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java [hunk 1]
```diff
@@ -122,7 +121,7 @@ public void testCreateContextThrowsInvalidSortException() throws Exception {
 				_mockResource));
 	}
 
-	@Test(expected = NotAcceptableException.class)
+	@Test
 	public void testCreateContextThrowsNotAcceptable() throws Exception {
```
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java:125]: Do you mean to delete it?

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java [hunk 2]
```diff
@@ -130,7 +129,7 @@ public void testCreateContextThrowsInvalidFilterException()
 				_mockResource));
 	}
 
-	@Test(expected = NotAcceptableException.class)
```
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java:133]: Do you mean to delete it?

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java:125]: I am asking about the purpose of this test

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java:133]: I am asking about the purpose of this test

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java:125]: The purpose of this test is just to create a Filter/Sort context and throwing a `NotAcceptableException` as they were implementing the previous AcceptLanguage logic. I think we could delete it as both of them has got a test called `testCreateContext`. What do you think?

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java [hunk 3]
```diff
@@ -82,9 +81,7 @@ else if (acceptLanguage.equals("zh-Hant-TW")) {
 				});
 
 			if (ListUtil.isEmpty(locales)) {
-				throw new NotAcceptableException(
-					"No locales match the accepted languages: " +
-						acceptLanguage);
+				return Collections.singletonList(company.getLocale());
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:84]: I am also reviewing this code:

```
public Locale getPreferredLocale() {
	List<Locale> locales = getLocales();

	if (ListUtil.isNotEmpty(locales)) {
		return locales.get(0);
	}

	try {
		User user = _portal.initUser(_httpServletRequest);

		return user.getLocale();
	}
	catch (NoSuchUserException noSuchUserException) {
		throw new NotFoundException(
			"Unable to get preferred locale from nonexistent user",
			noSuchUserException);
	}
	catch (Exception exception) {
		throw new InternalServerErrorException(
			"Unable to get preferred locale: " + exception.getMessage(),
			exception);
	}
}
```

Does it make sense this code now?

Comment (sergiojimcos) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:84]: Makes sense! I will take a view at this

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java [hunk 4]
```diff
@@ -179,18 +178,8 @@ private void _testCreateContext(Locale userLocale, User user)
...
-		}
-		catch (Exception exception) {
-			Assert.assertEquals(
-				NotAcceptableException.class, exception.getClass());
-			Assert.assertEquals(
-				"No locales match the accepted languages: es-es",
-				exception.getMessage());
-		}
+		Assert.assertEquals(
+			LocaleUtil.getDefault(), acceptLanguage.getPreferredLocale());
```
[the remark this answers is not in the document]
Comment (sergiojimcos) [reply modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java:182]: Is not the same as what we have in the `_testCreateContext` piece of code? :thinking:

**Comment by sergiojimcos** [comment]
ci:test:sf

**Comment by sergiojimcos** [comment]
ci:test:relevant

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java [hunk 5]
```diff
@@ -82,6 +84,13 @@ else if (acceptLanguage.equals("zh-Hant-TW")) {
 				});
 
 			if (ListUtil.isEmpty(locales)) {
+				if (Objects.equals(
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:87]: The if clause could be simplified. Please, look for the same pattern in the portal:

image

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java [hunk 6]
```diff
@@ -82,6 +84,13 @@ else if (acceptLanguage.equals("zh-Hant-TW")) {
...
+						Http.Method.valueOf(_httpServletRequest.getMethod()),
+						Http.Method.GET)) {
+
+					return Collections.emptyList();
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:91]: Does it make sense to perform an extra if clause and return an empty list if this code **is only executed when there is an empty list**?

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java [hunk 7]
```diff
@@ -104,41 +106,48 @@ public void testCreateContextWithAuthenticatedUser() throws Exception {
...
 
-		_testCreateContext(LocaleUtil.TAIWAN, user);
+		_testCreateContext(LocaleUtil.TAIWAN, Http.Method.GET, user);
+
+		_testCreateContext(LocaleUtil.TAIWAN, Http.Method.POST, user);
 	}
 
-	private void _testCreateContext(Locale userLocale, User user)
+	private void _testCreateContext(
+			Locale userLocale, Http.Method method, User user)
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java:124]: Not sorted alphabetically

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java [hunk 8]
```diff
@@ -210,7 +228,7 @@ private class AcceptLanguageMockHttpServletRequest
 		extends MockHttpServletRequest {
 
 		public AcceptLanguageMockHttpServletRequest(
-				User user, Locale... locales)
+				User user, Http.Method method, Locale... locales)
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/AcceptLanguageContextProviderTest.java:231]: method < user

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java [hunk 9]
```diff
@@ -122,27 +119,6 @@ public void testCreateContextThrowsInvalidSortException() throws Exception {
 				_mockResource));
 	}
 
-	@Test(expected = NotAcceptableException.class)
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/SortContextProviderTest.java:125]: In this case, as with the GET method, I think we should rewrite the test (I didn't know when I started reviewing the PR that for the methods != GET, an exception would not be thrown). Is it possible to rewrite the test to verify for the GET method that an exception is NOT thrown and for a different method that the exception is still being thrown?

File: modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java [hunk 10]
```diff
@@ -130,29 +127,6 @@ public void testCreateContextThrowsInvalidFilterException()
...
```
Comment (ccorreagg) [inline modules/apps/portal-vulcan/portal-vulcan-test/src/testIntegration/java/com/liferay/portal/vulcan/internal/jaxrs/context/provider/test/FilterContextProviderTest.java:133]: Same here

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please review my comments

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java (hunk 6 above)
Comment (ccorreagg) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/accept/language/AcceptLanguageImpl.java:91]: I mean, should not be better to include only one if clause with both conditions?

**Comment by sergiojimcos** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:forward

**Review by ccorreagg — APPROVED** [review]

**Comment by sergiojimcos** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:report:2554039347

**Comment by ccorreagg** [comment]
ci:forward:force

**Comment by ccorreagg** [comment]
ci:forward:force

**Pull request closed without merging**